subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy

```diff
@@ -325,4 +329,220 @@ project(':app') {
+        cantBeDownloaded.pom.allowGetOrHead()
+        cantBeDownloaded.artifact.expectDownloadBroken()
+
+        buildFile << """
+            project(':app') {
+                task resolve(type: Copy) {
+                    def artifacts = configurations.implementation.incoming.artifactView {
+                        attributes { it.attribute(artifactType, 'size') }
+                    }.artifacts
+                    from artifacts.artifactFiles
+                    into "\${buildDir}/libs"
+                }
+            }        
+            
+            project(':lib') {
+                dependencies {
+                    implementation "test:cant-be-downloaded:4.3"
+                }
+            }
+        """
+
+        when:
+        fails "resolve"
+
+        then:
+        output.count('Transforming') == 4
+        failure.assertResolutionFailure(":app:implementation")
+        failure.assertThatCause(Matchers.containsString("Could not download cant-be-downloaded.jar (test:cant-be-downloaded:4.3)"))
+        output.contains("Single step transform received dependencies files [] for processing common.jar")
+        output.contains("Single step transform received dependencies files [] for processing slf4j-api-1.7.25.jar")
+        output.contains("Single step transform received dependencies files [] for processing hamcrest-core-1.3.jar")
+        output.contains("Single step transform received dependencies files [hamcrest-core-1.3.jar] for processing junit-4.11.jar")
+    }
+
+    def "transform does not execute when dependencies cannot be transformed"() {
+        given:
+        buildFile << """
+            project(':app') {
+                task resolve(type: Copy) {
+                    def artifacts = configurations.implementation.incoming.artifactView {
+                        attributes { it.attribute(artifactType, 'end') }
+                    }.artifacts
+                    from artifacts.artifactFiles
+                    into "\${buildDir}/libs"
+                }
+            }        
+        """
+
+        when:
+        fails "resolve", '-DfailTransformOf=slf4j-api-1.7.25.jar'
+
+        then:
+        output.count('Transforming') == 8
+        failure.assertResolutionFailure(":app:implementation")
+        failure.assertThatCause(Matchers.containsString("Failed to transform file 'slf4j-api-1.7.25.jar'"))
+        ['lib.jar', 'common.jar', 'hamcrest-core-1.3.jar', 'junit-4.11.jar', 'slf4j-api-1.7.25.jar'].each {
+            assert output.contains("Transforming without dependencies ${it} to ${it}.txt")
+        }
+        [
+                'common.jar': [],
+                'hamcrest-core-1.3.jar': [],
+                'junit-4.11.jar': ['hamcrest-core-1.3.jar'],
+        ].each { artifact, dependencies ->
+            assert output.contains("Transform step 2 received dependencies files [${dependencies.collect { it + '.txt'}.join(', ')}] for processing ${artifact}.txt")
+        }
```

Comment:
The iteration doesn't seem to add anything. Either turn this into a method that takes the `Map`, or just write out each `output.contains ...` line. Either solution should be easier to understand than duplicating this logic in several places.